// Code generated by github.com/tizz98/results; DO NOT EDIT.
package results

import (
	"context"
	"fmt"
)

type StringResult struct {
	value *string
	err   error
}

// SetNewStringResult is a shortcut to creating a new StringResult and then calling .Set(v, err) on it.
func SetNewStringResult(v string, err error) (result StringResult) {
	result.Set(v, err)
	return
}

// SetNewStringResultPtr is a shortcut to creating a new StringResult and then calling .Set(v, err) on it.
// This function differs from SetNewStringResult by returning a pointer to StringResult.
func SetNewStringResultPtr(v string, err error) *StringResult {
	result := SetNewStringResult(v, err)
	return &result
}

// NewOptionalStringResult is a shortcut to creating a new StringResult and then calling .SetOptional(v, err) on it.
func NewOptionalStringResult(v *string, err error) (result StringResult) {
	result.SetOptional(v, err)
	return
}

// NewOptionalStringResultPtr is a shortcut to creating a new StringResult and then calling .SetOptional(v, err) on it.
// This function differs from NewOptionalStringResult by returning a pointer to StringResult.
func NewOptionalStringResultPtr(v *string, err error) *StringResult {
	result := NewOptionalStringResult(v, err)
	return &result
}

// IsOk returns true when the result contains a non-nil result with no error
func (r StringResult) IsOk() bool {
	return r.err == nil
}

// IsErr returns true when the result contains a non-nil error
func (r StringResult) IsErr() bool {
	return r.err != nil
}

// Unwrap panics if the result contains an error, otherwise it returns the value
func (r StringResult) Unwrap() string {
	if r.IsErr() {
		panic("cannot unwrap StringResult, it is an error")
	}
	return *r.value
}

// UnwrapTo will call the .Err() method on the other Result if this StringResult has an error.
// If other is a pointer to a StringResult, then .Ok() will be called if this StringResult name does not have an error.
func (r StringResult) UnwrapTo(other Result) Result {
	if r.IsErr() {
		other.Err(r.GetErr())
	} else if other, ok := other.(*StringResult); ok {
		other.Ok(r.Unwrap())
	}

	return other
}

// Expect panics with the specified message if the result contains an error, otherwise it returns the value
func (r StringResult) Expect(message string) string {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", message, r.GetErr()))
	}
	return *r.value
}

// Expectf panics with the specified message if the result contains an error, otherwise it returns the value.
// This is different than Expect because if will automatically format the string with the given args.
func (r StringResult) Expectf(format string, args ...interface{}) string {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", fmt.Sprintf(format, args...), r.GetErr()))
	}
	return *r.value
}

// UnwrapOr returns the value if there is not an error, otherwise the specified value is returned
func (r StringResult) UnwrapOr(v string) string {
	if r.IsOk() {
		return r.Unwrap()
	}
	return v
}

// UnwrapOrElse returns the value if there is not an error, otherwise the function is called and the result is returned
func (r StringResult) UnwrapOrElse(fn func(err error) string) string {
	if r.IsOk() {
		return r.Unwrap()
	}
	return fn(r.err)
}

// Ok sets the result to a successful result with the provided value.
// This will panic if the result has already been set to successful or an error.
func (r *StringResult) Ok(v string) {
	r.clear()
	r.value = &v
}

// Err sets the result to an error result with the provided error.
// This will panic if the result has already been set to successful or an error.
func (r *StringResult) Err(err error) {
	r.clear()
	r.err = err
}

// GetError returns the error of the result. It may be nil, so check with StringResult.IsErr() first.
func (r StringResult) GetErr() error {
	return r.err
}

// Tup returns a tuple of (string, error) with "" being returned for string if there is an error
func (r StringResult) Tup() (string, error) {
	return r.UnwrapOr(""), r.err
}

// Set is a shortcut to checking the value of an error before setting the result.
// If there is an error, StringResult.Err(err) will be called, otherwise StringResult.Ok(v) will be called.
func (r *StringResult) Set(v string, err error) {
	if err != nil {
		r.Err(err)
	} else {
		r.Ok(v)
	}
}

// SetOptional is similar to Set but can be called with a nil value for string.
// The error will be checked first to see if it is not nil, then if v is not nil it will be set with .Ok(v).
func (r *StringResult) SetOptional(v *string, err error) {
	if err != nil {
		r.Err(err)
	} else if v != nil {
		r.Ok(*v)
	}
}

func (r StringResult) clear() {
	r.value = nil
	r.err = nil
}

// ResultToStringResult takes a Result interface and returns a StringResult. If "r" contains a StringResult,
// it is returned, otherwise a new StringResult is returned with an error set.
func ResultToStringResult(r Result) (result StringResult) {
	v, ok := r.(*StringResult)
	if !ok {
		result.Err(fmt.Errorf("expected *StringResult got %T", v))
		return
	}

	return *v
}

// ContextWithString embeds the given value of string into the context for later retrieval with StringFromContext
func ContextWithString(ctx context.Context, key interface{}, v string) context.Context {
	return context.WithValue(ctx, key, v)
}

// StringFromContext attempts to retrieve a string value from the specified context. A StringResult is returned
// which can be used to inspect the success or failure of retrieval.
func StringFromContext(ctx context.Context, key interface{}) (result StringResult) {
	if v, ok := ctx.Value(key).(string); !ok {
		result.Err(fmt.Errorf("%#v not found in context", key))
	} else {
		result.Ok(v)
	}
	return
}
